java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowBufReadHolder.java

```diff
@@ -0,0 +1,29 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.ipc.message;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Extend MessageReadHolderBase to contain a message body ArrowBuf used when reading messages
+ * with MessageReader.
+ */
+public class ArrowBufReadHolder extends MessageReadHolder {
```

Comment:
I don't follow having two different holder implementations. The message should always be in ArrowBufs (see comment on the base MessageReadHolder below).

Reply:
This adds an `ArrowBuf` to hold the message body off heap . In my use case, I don't need to load the body to an `ArrowBuf` and doing so would just add extra overhead in copying the data back and forth.  It would be possible to combine these into one holder class, because most people will probably read the message body off heap. In my case, I would just not set this value in the holder.  Does that sound better?  

For example:
```java
public class MessageReadHolder {
  public int messageLength;
  public ByteBuffer messageBuffer;
  public Message message;
  public ArrowBuf bodyBuffer;
}
```